### PR TITLE
Fix for #14

### PR DESCRIPTION
**Pull request by JocelynVelarde**
## Changes made
1. Updated `BNO.cpp` and library to calculate target angle for robot movement
2. Created new library `Uart.cpp` to organize serial communication in between robot camera and control movement. This library contains the following relevant methods: `update()` - `getDistance()` - `getAngle()`

Note that the `target_angle` is calculated using the following model

`BNO_angle - ball_angle = difference_angle`
`2pi - difference_angle = target_angle`

## Issue ticket number and link
Refers to issue #14

### CODE REVIEW

File: Control/Uart.cpp [hunk 1]
```diff
@@ -0,0 +1,33 @@
+#include "Arduino.h"
+#include "Uart.h"
+
+Uart::Uart() : stringComplete(false) {
+}
+
+void Uart::initializeCommUart() {
+    Serial.begin(9600);
+}
```
Comment (afr2903) [inline Control/Uart.cpp:9]: I was checking this [C++ naming convention of Google](https://google.github.io/styleguide/cppguide.html#Naming), in summary:

- Functions have to be camel case with capital at the beginning: `IntializeComm`.
- Variables have to be snake case: `in_char`

File: Control/Uart.h [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    void initializeCommUart();
+    void update();
+    float getDistance();
+    float getAngle();
+
+private:
+    String inputString;
+    bool stringComplete;
+    float distance;
+    float angle;
```
Comment (afr2903) [inline Control/Uart.h:18]: According to the convention, this member variables should be snake case with a final underscore: `input_string_`

File: Sensors/BNO/Bno.cpp [hunk 3]
```diff
@@ -20,8 +24,12 @@ void BNO055::InitializeBNO() {
 }
 
 void BNO055::getBNOData() {
+    Uart uart;
+    uart.update();
     imu::Vector<3> euler = bno.getVector(Adafruit_BNO055::VECTOR_EULER);
```
Comment (afr2903) [inline Sensors/BNO/Bno.cpp:29]: Make a single instance of the Uart object and intialize it in the constructor of intialization method. This code right now creates a new instance of `Uart` each time BNO Data is required.

File: Pico/Motors/Motors.cpp [hunk 4]
```diff
@@ -65,6 +65,22 @@ void Motors::moveForward() {
...
+};
+
+void Motors::moveLeft() {
+    stopMotors();
+    motor1.moveBackward();
+    motor2.moveForward();
+    motor3.moveForward();
+    motor4.moveBackward();
+};
+
```
Comment (afr2903) [inline Pico/Motors/Motors.cpp:83]: I think you should remove all `moveDirection()` functions and instead have a general `Move(angle)` method where you enter the desired angle and calculate the forward kinematics of each motor.

Also, the `stopMotors()` function at the beginning should be used for hard stops, and the internal management of the motors speed could benefit from a speed ramp instead.

File: Pico/mainRP.ino [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+uint8_t motor3In2 = 11;
+uint8_t motor3PWM = 21;
+
+uint8_t motorIn1 = 8;
+uint8_t motorIn2 = 9;
+uint8_t motorPWM = 10;
+
+uint8_t motor2In1 = 22;
+uint8_t motor2In2 = 23;
+uint8_t motor2PWM = 15;
```
Comment (afr2903) [inline Pico/mainRP.ino:22]: The information of the PINS could be added to a `Constants.h` file with an static class for easier data handling.

**Review by afr2903 — CHANGES_REQUESTED** [review]
There are another improvements that could be useful but for the upcoming deadline, I think we should review them later, for example the synchronization and sample times for obtaining sensors data could be useful in the future when the code gets more complex. Right now, the tests and control tuning are our main goals :)

File: Control/Uart.cpp (hunk 1 above)
Comment (JocelynVelarde) [reply Control/Uart.cpp]: Sure, I will create a new issue and link a branch to fix convention standards while I am away.

File: Sensors/BNO/Bno.cpp (hunk 3 above)
Comment (JocelynVelarde) [reply Sensors/BNO/Bno.cpp:29]: Already fixed this, it will be reflected on my new PR, thanks

File: Pico/Motors/Motors.cpp (hunk 4 above)
Comment (JocelynVelarde) [reply Pico/Motors/Motors.cpp:83]: I do have a method called `moveMotors()` that receives as parameters the angle and PWM speed, should I apply it inside this specific logic? 

In regard to the `stopMotors()` method, would you rather prefer that motor speed is decreased proportional to the stop command?

File: Pico/mainRP.ino (hunk 5 above)
Comment (JocelynVelarde) [reply Pico/mainRP.ino:22]: Great! This will also be fixed while I am away on a new issue

**Comment by JocelynVelarde** [comment]
PR closed because implementation already has been taken care of in #25

**Pull request closed without merging**